tests/rptest/tests/compatibility/flink_test.py

```diff
@@ -0,0 +1,122 @@
+# Copyright 2021 Vectorized, Inc.
+#
+# Use of this software is governed by the Business Source License
+# included in the file licenses/BSL.md
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0
+
+from ducktape.mark.resource import cluster
+from ducktape.utils.util import wait_until
+
+from rptest.services.compatibility.example_runner import ExampleRunner
+from rptest.services.compatibility.flink_cluster import FlinkCluster
+import rptest.services.compatibility.flink_examples as FlinkExamples
+from rptest.services.kaf_producer import KafProducer
+from rptest.services.rpk_consumer import RpkConsumer
+from rptest.tests.redpanda_test import RedpandaTest
+from rptest.clients.types import TopicSpec
+import math
+
+
+class FlinkTest(RedpandaTest):
+    """
+    Base class for Flink tests that contains all
+    shared objects between tests and starts the
+    Flink cluster
+    """
+
+    # The JOB is the java program that uses Flink with kafka connectors.
+    # The java program is represented by a wrapper in FlinkExamples.
+    JOB = None
+
+    # The producer should be an extension to KafProducer
+    PRODUCER = None
+
+    def __init__(self, test_context):
+        super(FlinkTest, self).__init__(test_context=test_context)
+
+        self._ctx = test_context
+
+        self._max_records = 10
+        self._timeout = 120
+
+        self._flink_cluster = FlinkCluster(test_context)
+
+    def is_valid_msg(self, msg):
+        raise NotImplementedError("is_valid_msg() is undefined")
+
+    def setUp(self):
+        self._flink_cluster.start()
+
+        super().setUp()
+
+    @cluster(num_nodes=6)
+    def test_flink(self):
+        # This will raise TypeError if JOB is undefined
+        flink_job = self.JOB(self.redpanda)
+
+        # This will raise TypeError if PRODUCER is undefined
+        producer = self.PRODUCER(self._ctx, self.redpanda, self.topics[0].name,
+                                 self._max_records)
+        consumer = RpkConsumer(self._ctx, self.redpanda, self.topics[1].name)
+
+        # Start the flink job on the
+        # cluster node
+        self._flink_cluster.run_flink_job(flink_job, self._timeout)
+
+        # Produce some messages
+        producer.start()
+        producer.wait()
+
+        # Consume the data
+        consumer.start()
+
+        def try_cons():
+            i = 0
+            msgs = consumer.messages
+            while i < len(msgs) and not self.is_valid_msg(msgs[i]):
```

Comment:
I'm  bit confused by this loop: we drop out if the message is valid but only continue to iterate if the message is invalid?

Also it seems like somewhere in here we should be comparing the number of messages received with _max_records as a success condition.

Reply:
Ah.  So I think this can be made a bit more obvious by returning `consumer.messages and self.is_valid_msg(consumer.messages[-1])` with maybe a comment to explain that the example is going to output a bunch of messages with lower counts but we only care about it getting to the right count eventually.